src/Middleware/WebSockets/src/WebSocketMiddleware.cs

```diff
@@ -165,6 +155,70 @@ public async Task<WebSocket> AcceptAsync(WebSocketAcceptContext acceptContext)
 
                 return WebSocket.CreateFromStream(opaqueTransport, isServer: true, subProtocol: subProtocol, keepAliveInterval: keepAliveInterval);
             }
+
+            public static bool CheckSupportedWebSocketRequest(string method, IHeaderDictionary requestHeaders)
+            {
+                bool validUpgrade = false, validConnection = false, validKey = false, validVersion = false;
+
+                if (!string.Equals("GET", method, StringComparison.OrdinalIgnoreCase))
+                {
+                    return false;
+                }
+
+                var interestingHeaders = new List<KeyValuePair<string, string>>();
+                foreach (var headerName in HandshakeHelpers.NeededHeaders)
+                {
+                    foreach (var value in requestHeaders.GetCommaSeparatedValues(headerName))
+                    {
+                        interestingHeaders.Add(new KeyValuePair<string, string>(headerName, value));
+                    }
+                }
+
+                foreach (var pair in requestHeaders)
```

Comment:
Most of these loops can be removed and replaced by directly looking up the individual headers. Something like this:
```
var connection = requestHeaders.GetCommaSeparatedValues(HeaderNames.Connection);
if (!connection .Contains(HeaderNames.Upgrade, StringComparison.OrdinalIgnoreCase))
{
    return false;
}
if (connection.Count == 1)
{
    requestHeaders[HeaderNames.Connection] = HeaderNames.Upgrade;
}
```

Reply:
Getting below error when doing `connection.Contains(HeaderNames.Upgrade, StringComparison.OrdinalIgnoreCase)`

https://sharplab.io/#v2:C4LgTgrgdgPgAgJgAwFgBQiCM65IARyYB0AMgJZQCOA3OjpkgNoC6eAagM554C8eA3ngBEAQSEAaYQCEheAL6009AJwAKTkQDCAeyjAAhhQ6rREvAGVgYCgHMdAWwAO+6x11EAklABuLsvr1NCAAbYAgwAFMPGyhtSM19DgiASmTqIA=

Reply:
```
requestHeaders[HeaderNames.Connection].Contains(HeaderNames.Upgrade)
```

Would probably work, although you can't specify the `StringComparison`, and it's case sensitive I believe. For now you can just loop over the array of values and check them. But we might want to add the `StringComparison` overload to `StringValues.Contains` so we can avoid the `string[]` allocation.

Reply:
I was referring to this Contains method:
https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.contains?view=net-5.0#System_Linq_Enumerable_Contains__1_System_Collections_Generic_IEnumerable___0____0_System_Collections_Generic_IEqualityComparer___0__

You've also skipped the GetCommaSeparatedValues step which is important to keep.

Reply:
Still allocates an unneeded `string[]`

Reply:
Yea, lets not do that 😄

Reply:
GetCommaSeparatedValues is necessary since headers may contain more than one value per line. We should discuss optimizations for GetCommaSeparatedValues in a separate issue.

Reply:
I mean don't use Contains.

Reply:
@Tratcher The namespace was already there! For the moment, I was also confused why it's not working!